### DISCUSSION TITLE
[Macrobenchmarks] - It is wrong to assume device.findObject() will always return the object.

### DISCUSSION THREAD

**Discussion by SunnyBe** [android/nowinandroid#711] [Q&A] [answered]
On my personal project as well as the nowinandroid project, I have experienced sometimes the baseline generator/macrobenchmark fails to find this object sometimes.

```
fun MacrobenchmarkScope.interestsScrollTopicsDownUp() {
        val topicsList = device.findObject(By.res("interests:topics"))
        device.flingElementDownUp(topicsList)
    }
```

Do we expect the test to fail and crash the app(stopping after the last iteration) or skip the process when the object is not found and continue with the tests.

Why not use this?
```
fun MacrobenchmarkScope.interestsScrollTopicsDownUp() {
          val topicsList = device.findObject(By.res("interests:topics"))
          if (topicsList != null) device.flingElementDownUp(topicsList)
     }
```

**Comment by SimonMarquis** [comment, marked as answer]
IMO, we should **never** silently ignore this kind of null return value. This is definitely a bad testing practice and will hide potential bugs.
Unless there is a very good reason to handle it, for instance if we want to check 2 potential different code paths, you probably never should do that.

Unfortunately, `UiAutomator` has been designed in Java, and never received proper nullability annotations.
Although the Javadoc explicitly mention the potential null return value:
> Returns the first object to match the selector criteria, or null if no matching objects are found.

This means that tests will fail with a semantically "wrong" `NullPointerException` on the next operation:

```kotlin
device.findObject(By.text("foo")).click()
```

```
java.lang.NullPointerException: Attempt to invoke virtual method 'void androidx.test.uiautomator.UiObject2.click()' on a null object reference
```

A better solution would be to create this kind of extension:

```kotlin
fun UiDevice.getObject(selector: BySelector): UiObject2 =
    findObject(selector) ?: error("Object not found for: $selector")
```

And using it like this:

```kotlin
device.getObject(By.text("foo")).click()
```

will produce this error message:

```
java.lang.IllegalStateException: Object not found for: BySelector [TEXT='\Qfoo\E']
```

---

In this project, we already have this kind of extension that behaves similarly `fun UiDevice.waitAndFindObject(selector: BySelector, timeout: Long): UiObject2`.

**Reply by gptshubham595** [reply]
Can you tell in By.text we have to pass android:id ?

like I have a button and a textview and a imageview?
